pkg/controllers/updaterun/validation.go

```diff
@@ -0,0 +1,319 @@
+	klog.V(2).InfoS("Start to validate the clusterStagedUpdateRun", "clusterStagedUpdateRun", updateRunRef)
+
+	// Validate the ClusterResourcePlacement object referenced by the ClusterStagedUpdateRun.
+	placementName, err := r.validateCRP(ctx, updateRunCopy)
+	if err != nil {
+		return -1, nil, nil, err
+	}
+	// Validate the applyStrategy.
+	if !reflect.DeepEqual(updateRun.Status.ApplyStrategy, updateRunCopy.Status.ApplyStrategy) {
+		mismatchErr := fmt.Errorf("the applyStrategy in the clusterStagedUpdateRun is outdated, latest: %v, recorded: %v", updateRunCopy.Status.ApplyStrategy, updateRun.Status.ApplyStrategy)
+		klog.ErrorS(mismatchErr, "the applyStrategy in the clusterResourcePlacement has changed", "clusterResourcePlacement", placementName, "clusterStagedUpdateRun", updateRunRef)
+		return -1, nil, nil, fmt.Errorf("%w: %s", errStagedUpdatedAborted, mismatchErr.Error())
+	}
+
+	// Retrieve the latest policy snapshot.
+	latestPolicySnapshot, clusterCount, err := r.determinePolicySnapshot(ctx, placementName, updateRunCopy)
+	if err != nil {
+		return -1, nil, nil, err
+	}
+	// Make sure the latestPolicySnapshot has not changed.
+	if updateRun.Status.PolicySnapshotIndexUsed != latestPolicySnapshot.Name {
+		mismatchErr := fmt.Errorf("the policy snapshot index used in the clusterStagedUpdateRun is outdated, latest: %s, recorded: %s", latestPolicySnapshot.Name, updateRun.Status.PolicySnapshotIndexUsed)
+		klog.ErrorS(mismatchErr, "there's a new latest policy snapshot", "clusterResourcePlacement", placementName, "clusterStagedUpdateRun", updateRunRef)
+		return -1, nil, nil, fmt.Errorf("%w: %s", errStagedUpdatedAborted, mismatchErr.Error())
+	}
+	// Make sure the cluster count in the policy snapshot has not changed.
+	if updateRun.Status.PolicyObservedClusterCount != clusterCount {
+		mismatchErr := fmt.Errorf("the cluster count initialized in the clusterStagedUpdateRun is outdated, latest: %d, recorded: %d", clusterCount, updateRun.Status.PolicyObservedClusterCount)
+		klog.ErrorS(mismatchErr, "the cluster count in the policy snapshot has changed", "clusterResourcePlacement", placementName, "clusterStagedUpdateRun", updateRunRef)
+		return -1, nil, nil, fmt.Errorf("%w: %s", errStagedUpdatedAborted, mismatchErr.Error())
+	}
+
+	// Collect the clusters by the corresponding ClusterResourcePlacement with the latest policy snapshot.
+	scheduledBindings, toBeDeletedBindings, err := r.collectScheduledClusters(ctx, placementName, latestPolicySnapshot, updateRunCopy)
+	if err != nil {
+		return -1, nil, nil, err
+	}
+
+	// if condition.IsConditionStatusFalse(meta.FindStatusCondition(updateRun.Status.Conditions, string(placementv1alpha1.StagedUpdateRunConditionProgressing)), updateRun.Generation) {
+	// 	// The clusterStagedUpdateRun has not started yet.
+	// 	klog.V(2).InfoS("Starting the staged update run from the beginning", "clusterStagedUpdateRun", updateRunRef)
+	// 	return 0, scheduledBindings, toBeDeletedBindings, nil
+	// }
```

Comment:
why comment it out?

Reply:
I actually want to remove this. With this shortcut, we are not able to verify if there's any mismatch in the stageStatus. This case should be covered in code already. If no stage has started, `updatingStageIndex == lastFinishedStageIndex == -1`.
```
// If no stage is updating, continue from the last finished stage (which will result it starting from 0).
   if updatingStageIndex == -1 {
       updatingStageIndex = lastFinishedStageIndex + 1
   }
```

Reply:
it's a bit different though. The condition is a short cut to save any validation. Do we want that?

Reply:
I think it still worth going through the validation to make sure the clusters in the stages are not changed. It should be a rare case to hit this code though: Only when initialization is completed and then updating the updateRun as started failed.